File: item-list/src/test/java/com/gurunars/item_list/AnimalItem.java
package com.gurunars.item_list;


class AnimalItem implements Item {

    private final long id;
    private int version;
    private Type type;

    @Override
    public long getId() {
        return id;
    }

    @Override
    public Enum getType() {
        return type;
    }

    enum Type {
        MONKEY, TIGER, WOLF, LION
    }

    void update() {
        this.version++;
    }

    AnimalItem(long id, int version, Type type) {
        this.version = version;
        this.type = type;
        this.id = id;
    }

    AnimalItem(long id, int version) {
        this(id, version, Type.MONKEY);
    }

    AnimalItem(long id, Type type) {
        this(id, 0, type);
    }

    @Override
    public String toString() {
        return "#" + id + "{" + type + " @ " + version + "}";
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof AnimalItem) {
            AnimalItem other = (AnimalItem) obj;
            return version == other.version && type == other.type;
        }
        return false;
    }

    @Override
    public int hashCode() {
        return Long.valueOf(id).hashCode();
    }
}

File: item-list/src/test/kotlin/com/gurunars/item_list/AnimalItem.kt
package com.gurunars.item_list

internal data class AnimalItem(private val id: Long, private val version: Int, private val type: AnimalItem.Type) : Item {

    constructor(id: Long, type: AnimalItem.Type) : this(id,0,type)

    override fun getType(): Enum<*> {
        return type
    }

    override fun getId(): Long {
        return id
    }

    override fun payloadsEqual(other: Item): Boolean {
        return other is AnimalItem && version == other.version
    }

    internal enum class Type {
        MONKEY, TIGER, WOLF, LION
    }

    override fun toString(): String {
        return "#$id{$type @ $version}"
    }

}